### PR TITLE
Companies own its product sequence

### PR DESCRIPTION
**Pull request by TimLai125**
@yostashiro  Please check the changes and let me know if there is any gap between my ideas and yours.

### CODE REVIEW

File: product_internal_reference_code/models/res_company.py [hunk 1]
```diff
@@ -2,13 +2,57 @@
...
     _inherit = "res.company"
 
     internal_code_prefix = fields.Char(
         string="Internal Reference Prefix",
-        copy=True,
+        copy=False,
+        require=True,
+        help="A prefix code which is used to automatically generate the "
+             "internal reference when creating a new product."
     )
```
Comment (yostashiro) [inline product_internal_reference_code/models/res_company.py]: I prefer not to have this field.

File: product_internal_reference_code/models/res_company.py [hunk 2]
```diff
@@ -2,13 +2,57 @@
...
+        seq = {
+            'name': res.name + ' - Product Sequence',
+            'code': 'product.product.internal_code',
+            'prefix': prefix or '',
+            'padding': 5,
+            'number_increment': 1,
+            'company_id': res.id,
+        }
+        self.env['ir.sequence'].create(seq)
+        return res
```
Comment (yostashiro) [inline product_internal_reference_code/models/res_company.py]: Not sure if we want this method.  If we keep this, set `res.id` to prefix.  User can later change the value in ir.sequence view.

File: product_internal_reference_code/models/res_company.py [hunk 3]
```diff
@@ -2,13 +2,57 @@
...
+                seq = {
+                    'name': rc.name + ' - Product Sequence',
+                    'code': 'product.product.internal_code',
+                    'prefix': vals['internal_code_prefix'],
+                    'padding': 5,
+                    'number_increment': 1,
+                    'company_id': rc.id,
+                }
+                self.env['ir.sequence'].create(seq)
+        return super(ResCompany, self).write(vals)
```
Comment (yostashiro) [inline product_internal_reference_code/models/res_company.py]: No need to have this method.

File: product_internal_reference_code/models/product_product.py [hunk 4]
```diff
@@ -11,9 +11,24 @@ class ProductProduct(models.Model):
...
+                    'prefix': self.env.user.company_id.internal_code_prefix
+                              or '',
+                    'padding': 5,
+                    'number_increment': 1,
+                    'company_id': self.env.user.company_id.id
+                }
+                self.env['ir.sequence'].create(seq)
+                seq_code = self.env['ir.sequence'].with_context(
+                    force_company=self.env.user.company_id.id).next_by_code(
+                    'product.product.internal_code')
```
Comment (yostashiro) [inline product_internal_reference_code/models/product_product.py]: In what case there is no seq_code if ir.sequence record is generated when company is created?

Comment (TimLai125) [reply product_internal_reference_code/models/product_product.py]: The first company?

Comment (yostashiro) [reply product_internal_reference_code/models/product_product.py]: In that case, isn't creating an ir.sequence record for each company at installation of the module more appropriate?

File: product_internal_reference_code/models/res_company.py (hunk 1 above)
Comment (TimLai125) [reply product_internal_reference_code/models/res_company.py]: If this field is removed, users have to activate the debug mode to change the prefix directly (Settings > Technical / Sequences & Identifiers / Sequences)

File: product_internal_reference_code/models/product_product.py (hunk 4 above)
Comment (TimLai125) [reply product_internal_reference_code/models/product_product.py]: You mean having a post_hook method to create the ir.sequence?

File: product_internal_reference_code/models/res_company.py (hunk 1 above)
Comment (yostashiro) [reply product_internal_reference_code/models/res_company.py]: I think that's fine as part of initial configuration.  Besides the risk of not making the adjustment in ir.sequence record is negligible.

We want to keep the changes to the system minimal for now as the requirements might later be changed.

File: product_internal_reference_code/models/product_product.py (hunk 4 above)
Comment (yostashiro) [reply product_internal_reference_code/models/product_product.py]: Yes

File: product_internal_reference_code/models/res_company.py (hunk 2 above)
Comment (TimLai125) [reply product_internal_reference_code/models/res_company.py]: Lets keep this method to ensure there will be always an ir.sequence for each company

**Comment by TimLai125** [comment]
@yostashiro Please check the updates, the module is simplified.

**Pull request merged**